### PR TITLE
Use _cleanup_ for UnitFileList hash

### PR DESCRIPTION
**Pull request by dtardon**

### CODE REVIEW

File: src/systemctl/systemctl-list-unit-files.c [hunk 1]
```diff
@@ -143,27 +143,23 @@ int verb_list_unit_files(int argc, char *argv[], void *userdata) {
...
                 r = unit_file_get_list(arg_runtime_scope, arg_root, h, arg_states, strv_skip(argv, 1));
-                if (r < 0) {
-                        unit_file_list_free(h);
+                if (r < 0)
                         return log_error_errno(r, "Failed to get unit file list: %m");
-                }
 
                 n_units = hashmap_size(h);
 
                 units = new(UnitFileList, n_units ?: 1); /* avoid malloc(0) */
```
Comment (keszybz) [inline src/systemctl/systemctl-list-unit-files.c]: `new()` uses `malloc_multiply()` which avoids calling `malloc(0)`. Please drop this work-around (in a separate commit ofc.)

**Review by bluca — APPROVED** [review]

File: src/systemctl/systemctl-list-unit-files.c [hunk 2]
```diff
@@ -174,7 +170,6 @@ int verb_list_unit_files(int argc, char *argv[], void *userdata) {
                 }
```
Comment (keszybz) [inline src/systemctl/systemctl-list-unit-files.c:169]: Hmm.
```C
                        units[c++] = *u;
                        free(u);
```
This moves ownership of contents from `u` to `units`. After that, `hashmap_free(h)` was used which would not touch the values in the hashmap. But after your patch, the hashmap is freed using `unit_file_list_hash_ops`, which would try to deallocate the contents again. I think.

File: src/systemctl/systemctl-list-unit-files.c [hunk 3]
```diff
@@ -143,27 +143,23 @@ int verb_list_unit_files(int argc, char *argv[], void *userdata) {
...
 
                 units = new(UnitFileList, n_units ?: 1); /* avoid malloc(0) */
-                if (!units) {
-                        unit_file_list_free(h);
+                if (!units)
                         return log_oom();
-                }
 
                 HASHMAP_FOREACH(u, h) {
                         if (!output_show_unit_file(u, NULL, NULL))
```
Comment (keszybz) [inline src/systemctl/systemctl-list-unit-files.c:165]: This part was borked: `valgrind build/systemctl -t socket --root=/ list-unit-files >/dev/null` reports a massive leak. Please make sure that it works correctly with your patch, and mention that in the commit message.

File: src/shared/install.c [hunk 4]
```diff
@@ -3625,6 +3625,14 @@ Hashmap* unit_file_list_free(Hashmap *h) {
 
 DEFINE_TRIVIAL_CLEANUP_FUNC(UnitFileList*, unit_file_list_free_one);
 
+DEFINE_HASH_OPS_WITH_VALUE_DESTRUCTOR(
+        unit_file_list_hash_ops,
```
Comment (keszybz) [inline src/shared/install.c]: I think this should be called `unit_file_list_hash_ops_free`, because it frees the value.

**Review by keszybz — CHANGES_REQUESTED** [review]
Please add a test for `list-unit-files` with filtering that works on client side. I seems buggy.

File: src/shared/install.c (hunk 4 above)
Comment (dtardon) [reply src/shared/install.c]: Indeed. Changed.

File: src/systemctl/systemctl-list-unit-files.c (hunk 1 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-unit-files.c]: Done.

File: src/systemctl/systemctl-list-unit-files.c (hunk 2 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-unit-files.c:169]: Oops. Fixed.

File: src/systemctl/systemctl-list-unit-files.c (hunk 3 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-unit-files.c:165]: Right.

**Comment by DaanDeMeyer** [comment]
This still seems to be missing the test that @keszybz asked for?

**Comment by dtardon** [comment]
> Please add a test for `list-unit-files` with filtering that works on client side. I seems buggy.

Done.

File: src/systemctl/systemctl-list-unit-files.c [hunk 5]
```diff
@@ -157,8 +157,8 @@ int verb_list_unit_files(int argc, char *argv[], void *userdata) {
 
                 n_units = hashmap_size(h);
 
-                units = new(UnitFileList, n_units ?: 1); /* avoid malloc(0) */
-                if (!units)
+                units = new(UnitFileList, n_units);
+                if (!units && n_units > 0)
```
Comment (yuwata) [inline src/systemctl/systemctl-list-unit-files.c]: Actually we can ignore OOM error when `n_units == 0`, but even in that case, we usually call `log_oom()` and fail earlier. Please drop the added condition.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/shared/install.c [hunk 6]
```diff
@@ -3619,12 +3619,16 @@ static UnitFileList* unit_file_list_free_one(UnitFileList *f) {
...
-
 DEFINE_TRIVIAL_CLEANUP_FUNC(UnitFileList*, unit_file_list_free_one);
 
+DEFINE_HASH_OPS_WITH_VALUE_DESTRUCTOR(
+        unit_file_list_hash_ops_free,
+        char,
+        string_hash_func,
+        string_compare_func,
+        UnitFileList,
+        unit_file_list_free_one);
```
Comment (yuwata) [inline src/shared/install.c]: In the last commit, `unit_file_list_free()` is removed. Then, maybe better to rename `unit_file_list_free_one()` -> `unit_file_list_free()`.

File: src/systemctl/systemctl-list-unit-files.c (hunk 5 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-unit-files.c]: Done.

File: src/shared/install.c (hunk 6 above)
Comment (dtardon) [reply src/shared/install.c]: Yes, makes sense.

**Comment by dtardon** [comment]
> The sanitizer errors found by `CentOS CI (Arch Linux + sanitizers)` look related.

Yup. Nothing new, though, just discovered by the additional test coverage.

**Comment by poettering** [comment]
lgtm

**Pull request merged**